### PR TITLE
Vp/min 64 refactor insurance logic

### PR DESCRIPTION
**Pull request by dev-vadim**
**Description:**

Implement protocol interest transfer in on_finalized() in case if threshold is reached.
Rename insurance -> protocol_interest

**Changes in storage or API:**
Storage:

ControllerData:
- insurance_factor -> protocol_interest_factor
- added protocol_interest_threshold

Pool:
- total_insurance -> total_protocol_interest

API:

Controller:
- deposit_insurance -> deposit_interest
- redeem_insurance -> redeem_interest
- set_insurance_factor -> set_protocol_interest_factor
- added set_protocol_interest_threshold

Describe storage structure changes or API changes. Make sure to update the [API docs](https://minterestfinance.atlassian.net/wiki/spaces/MINTEREST/pages/134807557/Minterest+API).

### CODE REVIEW

File: runtime/src/tests.rs [hunk 1]
```diff
@@ -1495,3 +1495,93 @@ fn balance_liquidation_pools_two_pools_should_work_test() {
...
+		.pool_initial(CurrencyId::DOT)
+		.pool_initial(CurrencyId::ETH)
+		.build()
+		.execute_with(|| {
+			// Set price = 2.00 USD for all polls.
+			assert_ok!(set_oracle_price_for_all_pools(2));
+
+			// Set interest factor equal 0.75.
+			assert_ok!(Controller::set_protocol_interest_factor(
+				<Runtime as frame_system::Config>::Origin::root(),
```
Comment (lut1k) [inline runtime/src/tests.rs:1514]: Better to add a helper function origin_root ().
As on lines 340 - 346.

File: runtime/src/tests.rs [hunk 2]
```diff
@@ -1495,3 +1495,93 @@ fn balance_liquidation_pools_two_pools_should_work_test() {
...
+				LiquidityPools::pools(CurrencyId::DOT).total_protocol_interest,
+				441000000000000000u128
+			);
+
+			System::set_block_number(10000000);
+
+			assert_ok!(MinterestProtocol::repay(bob(), DOT, dollars(20_000)));
+			assert_eq!(pool_balance(DOT), dollars(80_000));
+
+			let total_protocol_interest: Balance = 3645120550951706945733;
```
Comment (lut1k) [inline runtime/src/tests.rs:1561]: `3_645_120_550_951_706_945_733` - it is more convenient to read

File: runtime/src/tests.rs [hunk 3]
```diff
@@ -1495,3 +1495,93 @@ fn balance_liquidation_pools_two_pools_should_work_test() {
 			assert_eq!(liquidation_pool_balance(CurrencyId::ETH), 60_000 * DOLLARS);
 		});
 }
+
+//------------ TEMPORARY Protocol interest transfer tests ----------------------
+
+// Protocol interest should be transferred to liquidation pool after block is finalized
+#[test]
+fn protocol_interest_transfer_should_work() {
```
Comment (lut1k) [inline runtime/src/tests.rs:1503]: You have implemented a good integration test. In my opinion, it is also necessary to add a unit test.
You can easily set the `total_protocol_interest parameter` for multiple pools when creating a test environment. Call the function on_finalize() and check the balances in the pools.
Check the case when there is not enough funds in the liquidity-pool and we send the entire balance to the liquidation-pool.

File: pallets/minterest-protocol/src/lib.rs [hunk 4]
```diff
@@ -40,21 +41,28 @@ pub mod module {
...
 
 		/// The origin which may call deposit/redeem/borrow/repay in Whitelist mode.
 		type WhitelistMembers: Contains<Self::AccountId>;
 
 		/// Weight information for the extrinsics.
 		type ProtocolWeightInfo: WeightInfo;
+
+		#[pallet::constant]
+		/// Maximum total borrow amount per pool in usd
+		type ProtocolInterestTransferThreshold: Get<Balance>;
```
Comment (lut1k) [inline pallets/minterest-protocol/src/lib.rs:65]: Am I correct in understanding that this setting will be the same for all pools?

File: pallets/minterest-protocol/src/lib.rs [hunk 5]
```diff
@@ -123,7 +131,37 @@ pub mod module {
...
+		fn on_finalize(_block_number: T::BlockNumber) {
+			T::EnabledCurrencyPair::get().iter().for_each(|currency_pair| {
+				let pool_id = currency_pair.underlying_id;
+				let total_protocol_interest = T::LiquidityPoolsManager::get_pool_total_protocol_interest(pool_id);
+				if total_protocol_interest < T::ProtocolInterestTransferThreshold::get() {
+					return;
+				}
+
+				let total_balance = T::ManagerLiquidityPools::get_pool_available_liquidity(pool_id);
+				let to_liquidation_pool = match total_balance.cmp(&total_protocol_interest) {
```
Comment (Apollo50) [inline pallets/minterest-protocol/src/lib.rs:144]: I think you can use method `min()` isntead matching.
let to_liquidation_pool = total_balance.min(total_protocol_interest);

File: runtime/src/tests.rs (hunk 1 above)
Comment (dev-vadim) [reply runtime/src/tests.rs]: Done

File: runtime/src/tests.rs (hunk 2 above)
Comment (dev-vadim) [reply runtime/src/tests.rs]: Done

File: pallets/minterest-protocol/src/lib.rs (hunk 4 above)
Comment (dev-vadim) [reply pallets/minterest-protocol/src/lib.rs]: Changed it to configured for every pool separately.

File: pallets/minterest-protocol/src/lib.rs (hunk 5 above)
Comment (dev-vadim) [reply pallets/minterest-protocol/src/lib.rs]: Done

File: runtime/src/tests.rs (hunk 3 above)
Comment (dev-vadim) [reply runtime/src/tests.rs:1509]: Added test for partial transfer, but without balance checking due to liquidity/liquidation pool address issue

**Review by lut1k — APPROVED** [review]

**Review by Apollo50 — DISMISSED** [review]

File: pallets/minterest-protocol/src/lib.rs [hunk 6]
```diff
@@ -123,7 +126,34 @@ pub mod module {
 	pub struct Pallet<T>(PhantomData<T>);
 
 	#[pallet::hooks]
-	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_finalize(_block_number: T::BlockNumber) {
```
Comment (kurotychb) [inline pallets/minterest-protocol/src/lib.rs:130]: Could you add a short description, what the goal of code below?

Is it possible to move this code into a separate function?

```
fn on_finalize(_block_number: T::BlockNumber) {
       call_your_function_that_do_something_great();
}
```

File: pallets/minterest-protocol/src/lib.rs [hunk 7]
```diff
@@ -123,7 +126,34 @@ pub mod module {
...
+				if let Some(new_protocol_interest) = total_protocol_interest.checked_sub(to_liquidation_pool) {
+					if T::MultiCurrency::transfer(
+						pool_id,
+						&T::ManagerLiquidityPools::pools_account_id(),
+						&T::ManagerLiquidationPools::pools_account_id(),
+						to_liquidation_pool,
+					)
+					.is_ok()
+					{
+						let _ = <LiquidityPools<T>>::set_pool_total_protocol_interest(pool_id, new_protocol_interest);
```
Comment (kurotychb) [inline pallets/minterest-protocol/src/lib.rs:151]: ```let _ =``` is redundant. Function returns nothing

```
pub fn set_pool_total_protocol_interest(pool_id: CurrencyId, new_total_protocol_interest: Balance) {
		Pools::<T>::mutate(pool_id, |r| r.total_protocol_interest = new_total_protocol_interest)
}
```

File: pallets/minterest-protocol/src/lib.rs [hunk 8]
```diff
@@ -123,7 +126,34 @@ pub mod module {
...
+				if total_protocol_interest < <Controller<T>>::controller_dates(pool_id).protocol_interest_threshold {
+					return;
+				}
+
+				let total_balance = T::ManagerLiquidityPools::get_pool_available_liquidity(pool_id);
+				let to_liquidation_pool = total_balance.min(total_protocol_interest);
+
+				// If no overflow and transfer is successful update pool state
+				if let Some(new_protocol_interest) = total_protocol_interest.checked_sub(to_liquidation_pool) {
+					if T::MultiCurrency::transfer(
```
Comment (kurotychb) [inline pallets/minterest-protocol/src/lib.rs:143]: Should we cover case if transfer fails? 
Warning, event, or something else?

Comment (Apollo50) [reply pallets/minterest-protocol/src/lib.rs]: We've already discussed that. We decided this implementation is rough and we are going to add an error/event handler in near future (*next task).

File: pallets/minterest-protocol/src/lib.rs (hunk 6 above)
Comment (dev-vadim) [reply pallets/minterest-protocol/src/lib.rs:130]: Moved pool processing in separate function

File: pallets/minterest-protocol/src/lib.rs (hunk 7 above)
Comment (dev-vadim) [reply pallets/minterest-protocol/src/lib.rs]: Done

**Review by lut1k — APPROVED** [review]

**Review by kurotychb — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**